### ISSUE TITLE
Another significant array bug, this time with changeRoutine

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#515]
changeRoutine does not work with getting arrays from another widget.  I have a demo room at https://virtualtabletop.io/r9d9.

The first timer has a changeRoutine to look for the id of the spinner.  It uses the old SELECT and GET methods.  Works fine.  But, if you try to use the timer changeRoutine to look for an array (use the variable options), the room freezes.

The second button is the same, except it uses the newer dynamic expression method.  Same result.

**Comment by robartsd** [comment]
Don't get anything in console. Are we triggering an infinite loop here?

**Comment by 96LawDawg** [comment]
I was just about to say I looked in console and didn't see anything either.  I don't know how it could be an infinite loop; nothing is changed.  Look for a variable that is not an array (id, x, y, etc.) everything is fine.  Look for something that is an array and it breaks.  The array doesn't change either.  Not sure how this would be a loop, but some bug in the array code.

**Comment by robartsd** [comment]
Nothing finishes changing, so the state never gets updated. My computer fan says something is changing.

**Comment by robartsd** [comment]
It works if I change it to millisecondsChangeRoutine it works. I think what is happening is that the property is being set to the array and that triggers changeRoutine. changeRoutine copies the array again, but because it is a deep copy, it is not the same array, so chageRoutine gets triggered again. ... infinite loop!

**Comment by robartsd** [comment]
Well, the GET function works for millisecondsChangeRoutine.

**Comment by robartsd** [comment]
We should probably perform a deep copy of value in stateManged.set() and compare JSON representations in the comparison to see if we are setting the same value.

**Comment by bjalder26** [comment]
I believe I tried this exact same thing.  I believe what is happening is that `SET`ting the array triggers the changeRoutine, so it is constantly being triggered.  That is why I switched to millisecondsChangeRoutine.

**Comment by 96LawDawg** [comment]
> I believe I tried this exact same thing. I believe what is happening is that `SET`ting the array triggers the changeRoutine, so it is constantly being triggered. That is why I switched to millisecondsChangeRoutine.

You are correct that this only happens when `SET` is included.

**Comment by robartsd** [comment]
I haven't tested #516 yet, but it should fix this.

**Comment by robartsd** [comment]
Can probably create a changeRoutine that increments a property in itself that would blow up as soon as any property of the widget changes. #516 will not fixt that case.

**Comment by 96LawDawg** [comment]
> Can probably create a changeRoutine that increments a property in the same widget that would blow up as soon as any property of the widget changes. #516 will not fixt that case.

Yes, that is definitely recursive and will form an infinite repeating loop.  But a (mostly) static array in another widget isn't that.  I tested the actual room I was working on when I found this problem and this change does fix it.  I'm going to approve it based on test, but recommend you wait for at least somebody with code knowledge to check your work before merging.
